In [3]:
import json, glob, boto3, os
import pdb
import pandas as pd

# Testing out json processing

In [4]:
session = boto3.Session()
BUCKET_NAME = 'snowbot-pv'

# S3 Connect
s3 = session.resource('s3')

bucket = s3.Bucket(BUCKET_NAME)

In [5]:
# parquet engines don't handle shifted timezones
TZ = pytz.timezone('America/Vancouver')

DATA_DIR = "../data/"
MERGED_FILENAME = "merged_file.json"
merged_file = DATA_DIR + MERGED_FILENAME

TEST_FILENAME = "test_file.json"
merged_test_file = DATA_DIR + TEST_FILENAME


In [6]:
# from https://alexwlchan.net/2019/07/listing-s3-keys/
def get_matching_s3_objects(bucket, prefix="", suffix=""):
    """
    Generate objects in an S3 bucket.

    :param bucket: Name of the S3 bucket.
    :param prefix: Only fetch objects whose key starts with
        this prefix (optional).
    :param suffix: Only fetch objects whose keys end with
        this suffix (optional).
    """
    s3 = boto3.client("s3")
    paginator = s3.get_paginator("list_objects_v2")

    kwargs = {'Bucket': bucket}

    # We can pass the prefix directly to the S3 API.  If the user has passed
    # a tuple or list of prefixes, we go through them one by one.
    if isinstance(prefix, str):
        prefixes = (prefix, )
    else:
        prefixes = prefix

    for key_prefix in prefixes:
        kwargs["Prefix"] = key_prefix

        for page in paginator.paginate(**kwargs):
            try:
                contents = page["Contents"]
            except KeyError:
                return

            for obj in contents:
                key = obj["Key"]
                if key.endswith(suffix):
                    yield obj


def get_matching_s3_keys(bucket, prefix="", suffix=""):
    """
    Generate the keys in an S3 bucket.

    :param bucket: Name of the S3 bucket.
    :param prefix: Only fetch keys that start with this prefix (optional).
    :param suffix: Only fetch keys that end with this suffix (optional).
    """
    for obj in get_matching_s3_objects(bucket, prefix, suffix):
        yield obj["Key"]


def merge_matching_jsons(save_file, suffix=""):

    result = []

    for f in get_matching_s3_keys(BUCKET_NAME, suffix=suffix):

        # Write the file from S3 into a local temp file
        with open('temp', 'wb') as tfw:
            bucket.download_fileobj(f, tfw)

        # Append the local temp file into the result list
        with open('temp', 'rb') as tfr:
            result.append(json.load(tfr))

    os.remove("temp")

    # Fill the output file with the merged content
    with open(save_file, "w") as outfile:
        json.dump(result, outfile)

# TBD: more efficient to go straight to df w/o saving json to file


def set_lifts_df_datatypes(df):

    # Important to set categories because when writing incrementally to parquet, some increments
    # may not include all statuses.  Manually setting the categories avoids errors due to
    # different catergory indexing between increments.
    status_cat_dtype = pd.api.types.CategoricalDtype(
        categories=['X', 'H', 'O'], ordered=True)

    # set datatypes for lift table
    df = df.astype({
        "liftID": 'category',
        "resortID": 'category',
        "liftName": 'category',
        "status": status_cat_dtype,
        "timeToRide": "int"
    })
    df["timestamp"] = pd.to_datetime(df["timestamp"])

    return df


def jsons_to_df(jsons, record_path):
    df = pd.DataFrame.from_dict(pd.json_normalize(
        jsons, record_path=record_path, meta='timestamp'))
    df = set_lifts_df_datatypes(df)
    return df


def load_merged_json_as_df(merged_file, record_path):
    # load the merged json as a dataframe
    with open(merged_file, "r") as f:
        d = json.load(f)
        df = jsons_to_df(d, record_path)
        return df


def get_status_changes(df, keep_oldest=False):
    """
    Filter out rows that do not represent a status change.

    Parameters
    ----------
    df : pandas.DataFrame
        Includes 'status' and timestamp columns.  Lists status of every lift for each timestamp.
    keep_oldest : boolean
        Indicates if the returned DataFrame should keep the oldest status for each lift even if
        a lift has no status changes.  This is so that the earliest status for each lift is not
        lost, and all lifts are listed the returned DataFrame even if their status has not
        changed.  Use `False` when there is just one DataFrame to process.  Use `True` is cases
        where the status chages will be appended to an existing dataframe that already has at
        least one row for each lift.

    Returns
    -------
    pandas.DataFrame
        Only includes the rows from the original dataframe where there was a change to a new
        status.
    """

    def calc_status_change(df, keep_oldest=keep_oldest):
        change_rows = df[df.status.ne(df.status.shift())]

        if keep_oldest:
            firstrow = df.loc[df['timestamp'].idxmin()]
            keep_df = firstrow.to_frame().T.append(change_rows)
        else:
            keep_df = change_rows

        # Remove so that we don't need to write another column to S3 as we scrape?
        # Just calculate it when plotting and predicting?
        # keep_df['time_diff'] = keep_df['timestamp'].diff(1).shift(-1)

        return keep_df

    df = df.groupby('liftName', group_keys=False)\
           .apply(calc_status_change)\
           .reset_index(drop=True)

    df = set_lifts_df_datatypes(df)

    return df

# TBD: may need to convert timestamp to days (e.g. for Tableau)

### Test code

In [ ]:
test_df = load_merged_json_as_df(merged_test_file, 'lifts')
test_df.sort_values(by=['liftID', 'timestamp'])

In [6]:
# TBD: test that len is 1 when keep_oldest=False and larger otherwise
get_status_changes(test_df, keep_oldest=False).sort_values(by=['liftID', 'timestamp'])

liftID resortID                 liftName status  timeToRide  \
0     69       13  Blackcomb Gondola Lower      O           7   

                         timestamp  
0 2020-01-03 00:19:09.631011-08:00

# Process lift json fies

In [7]:
merge_matching_jsons(suffix="lifts.json", save_file=merged_file)

In [8]:
lifts_df = load_merged_json_as_df(merged_file, 'lifts')

lifts_status_changes_df = get_status_changes(lifts_df, keep_oldest=True)

**NOTE:** `timeToRide` is just the time is takes to ride the lift, not the current wait time:

In [251]:
lifts_df.groupby("liftName")['timeToRide'].unique()

liftName
7th Heaven Express                 [6]
Big Red Express                    [8]
Blackcomb Gondola Lower            [7]
Blackcomb Gondola Upper            [7]
Catskinner Express                 [4]
Coca-Cola Tube Park                [4]
Creekside Gondola                  [7]
Crystal Ridge Express              [7]
Emerald 6 Express                  [6]
Excalibur Gondola Lower            [3]
Excalibur Gondola Upper            [5]
Excelerator Express                [6]
Fitzsimmons Express                [6]
Franz's Chair                      [8]
Garbanzo Express                   [7]
Glacier Express                    [6]
Harmony 6 Express                  [6]
Horstman T-Bar                     [4]
Jersey Cream Express               [5]
Magic Chair                        [6]
Olympic Chair                      [5]
Peak 2 Peak Gondola               [12]
Peak Express                       [3]
Showcase T-Bar                     [3]
Symphony Express                   [7]
T-Bars          

In [276]:
lifts_df

liftID resortID                 liftName status  timeToRide  \
0         69       13  Blackcomb Gondola Lower      X           7   
1         70       13  Blackcomb Gondola Upper      X           7   
2          5       13  Excalibur Gondola Lower      X           3   
3         71       13  Excalibur Gondola Upper      X           5   
4          8       13      Excelerator Express      X           6   
...      ...      ...                      ...    ...         ...   
84559     22       13      Peak 2 Peak Gondola      X          12   
84560     44       13            Franz's Chair      X           8   
84561     43       13             Peak Express      X           3   
84562     42       13         Symphony Express      X           7   
84563     41       13                   T-Bars      X           5   

                             timestamp  
0     2020-01-03 00:19:09.631011-08:00  
1     2020-01-03 00:19:09.631011-08:00  
2     2020-01-03 00:19:09.631011-08:00  
3     2020-01-03 00:19:09.631011-08:00  
4     2020-01-03 00:19:09.631011-08:00  
...                                ...  
84559 2020-02-03 10:30:26.557898-08:00  
84560 2020-02-03 10:30:26.557898-08:00  
84561 2020-02-03 10:30:26.557898-08:00  
84562 2020-02-03 10:30:26.557898-08:00  
84563 2020-02-03 10:30:26.557898-08:00  

[84564 rows x 6 columns]

In [9]:
def get_status_durations(lifts_df):
    '''Calculate values and add columns for the time difference between the
    timestamp for the current status and the timestamp for the next status
    for each lift:
    `time_diff` column: Gives the duration that the lift was in the status indicated in the `status` column.
    `time_diff_seconds` column: `time_diff` converted to seconds.
    
    lifts_status_changes_df should be TBD
    '''
    # TBD: optimize if needed via # 3 under:
    # https://towardsdatascience.com/pandas-tips-and-tricks-33bcc8a40bb9
    df = lifts_df.sort_values(by=['liftID', 'timestamp'])
    df['time_diff'] = df.groupby('liftID')['timestamp'].diff(1).shift(-1)

    # Fill in the durations which will be missing for the most recent status changes
    missing_time_diffs_idx = df.loc[(df['time_diff'].isnull()) & (
        df['timestamp'] >= df['timestamp'].min()), 'timestamp'].index.values

    df.loc[missing_time_diffs_idx, 'time_diff'] = df['timestamp'].max(
    ) - df.loc[missing_time_diffs_idx, 'timestamp']

    # Convert to seconds
    df['time_diff_seconds'] = df['time_diff'].dt.total_seconds()

    return df

In [10]:
df = get_status_durations(lifts_status_changes_df)

# Uses local date formatting, otherwise Tableau will mix up month and day
# alternatively, can export to json:
# lifts_status_changes_df.to_json(DATA_DIR + "lifts_status_changes.json", orient='table')
df.to_csv(DATA_DIR + "lifts_status_changes.csv", date_format='%c')


In [17]:
# add:
# 
# daily: for each chair calculate most open status of the day: O > H > X
# Days since each chair was last seen open with timestamp of most recent open time.
# snowfall since last open
# save data for other mountains

# Storage options testing

In [ ]:
df.to_pickle(DATA_DIR + "df_test.pkl")

In [ ]:
from fastparquet import write

# parquet engines don't handle shifted timezones
import pytz
TZ = pytz.timezone('America/Vancouver')
df['timestamp'] = df.timestamp.dt.tz_convert(pytz.utc)

In [ ]:
# Note: May need snappy-python as a req to run on AWS Lambda
df.to_parquet(DATA_DIR + "df_test.parquet", engine='fastparquet')

In [ ]:
load_df = pd.read_parquet(DATA_DIR + "df_test.parquet")
load_df['timestamp'] = load_df.timestamp.dt.tz_convert(TZ) # convert back to correct timezone


In [ ]:
#TBD convert back to correct datatypes
load_df.dtypes

In [ ]:
df.to_csv(DATA_DIR + "df_test.csv")

Test file size results:
- json: 800 Kb?
- csv: 474 Kb
- pickle: 145 Kb
- parquet: 15 Kb

## Delta Lake Testing

Requires apache spark instance.  For future use, could set one up to work with lambda using https://aws.amazon.com/emr/features/spark/?

Otherwise databricks (similar to QxMD project)

# json comparison and parquet to S3

In [18]:
from fastparquet import write, ParquetFile
import os
import pytz
import s3fs
import botocore

os.chdir("../src/data/snowbot_AWS_lambda/")
from scrape import get_data
os.chdir("../../../notebooks")

In [19]:
fs = s3fs.S3FileSystem()
myopen = fs.open
nop = lambda *args, **kwargs: None

HISTORY_FNAME = 'wb_lifts_history.parquet'
PRIOR_STATUS_FNAME = 'lifts_prior.json'

In [17]:
def write_dataframe_to_parquet_on_s3(df, fname):
    """ Write a dataframe to a Parquet file on S3.  Creates a new parquet file if one
    doesn't already exist.
    """

    def write_parquet(df, fname, app=True):

        output_file = f"s3://{BUCKET_NAME}/{fname}"
        write(output_file,
              df,
              # partition_on=['timestamp'],
              file_scheme='hive',
              append=app,  # need to remove or catch exception to work when file doesn't exist
              open_with=myopen,
              mkdirs=nop)
        print(f"Writing {len(df)} records to {fname}.")

    # Unshift the timezone because parquet engines don't handle shifted timezones
    df.loc[:, 'timestamp'] = df.loc[:, 'timestamp'].dt.tz_convert(pytz.utc)

    s3_object = s3.Object(BUCKET_NAME, fname)

    if not list(bucket.objects.filter(Prefix=fname)):
        print(f"File {fname} not found.  Creating new file.")
        # Keep oldest status for each lift because creating new file
        df = get_status_changes(df, keep_oldest=True)
        write_parquet(df, fname, app=False)

    else:
        print(f"File {fname} found on S3.")
        df = get_status_changes(df, keep_oldest=False)
        write_parquet(df, fname, app=True)

## Generalized version
For all data from the EpicMix API

In [13]:
import requests
from datetime import datetime

# Filter for a specific resort
def filter_resort(data, resortID=None):
    if resortID:
        return data["resortID"] == resortID
    else:
        return data

def get_data_all_TEST():
    """new version: Defaults to all resorts.  Otion to filter"""
    API_URL = 'http://www.epicmix.com/vailresorts/sites/epicmix/api/mobile/'
    DATA_LIST = {'lifts': 'lifts', 'weather': 'snowconditions', 'terrain': 'terrains'}  # keys are used in the requests, the values and used in the response
    json_data = dict()

    for d, name in DATA_LIST.items():
        res = requests.get(API_URL + d + '.ashx')
        res.raise_for_status()
        data = json.loads(res.text)[name]
        data = list(filter(filter_resort, data))
        json_data[d] = json.dumps({'timestamp': str(datetime.now(TZ)), d: data})

    return json_data

In [14]:
HISTORY_SUFFIX = '_history_DEV.parquet'
PRIOR_SUFFIX = '_prior_DEV.json'


class ParquetWriter():
    """Identifies new data and writes it to Parquet file on S3."""

    def __init__(self):
        # Get current data
        self.data_current_all = get_data_all_TEST()  # String.

    def write_new_data_all(self):
        """Writes new data for each type (i.e. 'lifts', 'weather', 'terrian')
        of data returned by the API.
        """
        # self.table is the type of data
        for self.table in self.data_current_all:
            self.current_json = json.loads(self.data_current_all[self.table])
            self.prior_fname = self.table + PRIOR_SUFFIX
            self.prior_object = s3.Object(BUCKET_NAME, self.prior_fname)
            self.write_new_data()

    def write_new_data(self):
        """If new data since the last update of Parquet file is found, add it to the Parquet
        file.  Save the current data as json to serve as the prior in the next comparison.
        """

        # Get prior data json
        try:
            self.prior_object.load()
        except botocore.exceptions.ClientError as e:
            if e.response['Error']['Code'] == "404":
                print("Prior doesn't exist")
                # Create the prior file
                self.save_prior_data()
                print(f"Created {self.prior_fname}")
            else:
                # Something else has gone wrong.
                raise
        else:
            # The prior data file exists
            self.get_prior_data()
            if self.data_changed():

                # Get a df with the status chages between the prior and current json data
                df = jsons_to_df([self.prior_json, self.current_json], record_path=self.table)
                write_dataframe_to_parquet_on_s3(df, self.table + HISTORY_SUFFIX)

                # save current data json as prior
                self.save_prior_data()
                print(
                    f"Replaced data in {self.prior_object.key} with current data.")

    def get_prior_data(self):
        prior = self.prior_object.get()[
            'Body'].read().decode('utf-8')
        self.prior_json = json.loads(prior)
        print(f"Loaded prior {self.table} json data from S3")

    def data_changed(self):
        """Compare current data json with prior data json without their timestamps.  The timestamps
        on the current json will always be more recent even when none of the other data has changed.
        """
        if self.prior_json[self.table] == self.current_json[self.table]:
            print(
                f"No differences between current and prior {self.table} data were found.")
            return False
        else:
            print(
                f"Found differences between current and prior {self.table} data.")
            return True

    def save_prior_data(self):
        """Save the current data as prior data on S3."""
        bucket.put_object(Key=self.prior_fname,
                          Body=bytes(json.dumps(self.current_json).encode('UTF-8')))

In [15]:
pr = ParquetWriter()
pr.write_new_data_all()

Loaded prior lifts json data from S3
Found differences between current and prior lifts data.


NameError: name 'write_dataframe_to_parquet_on_s3' is not defined

In [25]:
prior_object = s3.Object(BUCKET_NAME, 'weather' + PRIOR_SUFFIX)
prior = prior_object.get()[
            'Body'].read().decode('utf-8')
weather_prior_json = json.loads(prior)

prior_object = s3.Object(BUCKET_NAME, 'terrain' + PRIOR_SUFFIX)
prior = prior_object.get()[
            'Body'].read().decode('utf-8')
terrain_prior_json = json.loads(prior)

prior_object = s3.Object(BUCKET_NAME, 'lifts' + PRIOR_SUFFIX)
prior = prior_object.get()[
            'Body'].read().decode('utf-8')
lifts_prior_json = json.loads(prior)

In [27]:
weather_prior_json

{'timestamp': '2020-02-04 20:37:26.904928-08:00',
 'weather': [{'newSnow': '7',
   'last24Hours': '7',
   'last48Hours': '7',
   'last7Days': '13',
   'midMountainBase': '58',
   'resortID': 1,
   'weatherForecast': [{'daycode': 0,
     'dayDescription': 'Today',
     'forecastString': '',
     'iconName': 'partly-cloudy-night',
     'summaryDescription': 'Partly Cloudy',
     'temperatureHigh': '14',
     'temperatureLow': '-6'}]},
  {'newSnow': '7',
   'last24Hours': '7',
   'last48Hours': '7',
   'last7Days': '10',
   'midMountainBase': '49',
   'resortID': 2,
   'weatherForecast': [{'daycode': 0,
     'dayDescription': 'Today',
     'forecastString': '',
     'iconName': 'partly-cloudy-night',
     'summaryDescription': 'Mostly Cloudy',
     'temperatureHigh': '14',
     'temperatureLow': '-4'}]},
  {'newSnow': '3',
   'last24Hours': '5',
   'last48Hours': '5',
   'last7Days': '8',
   'midMountainBase': '52',
   'resortID': 3,
   'weatherForecast': [{'daycode': 0,
     'dayDescript

In [93]:
# TBD: check that weather_prior_json['weather'][0].keys() matches list of expected columns (in case new ones are in use)
# Add as exception handling for the other tables as well
weather_prior_json['weather'][0].keys()

dict_keys(['newSnow', 'last24Hours', 'last48Hours', 'last7Days', 'midMountainBase', 'resortID', 'weatherForecast'])

In [116]:
weather_meta_fields = ['newSnow', 'last24Hours', 'last48Hours', 'last7Days', 'midMountainBase', 'resortID']
weather_meta = [['weather', i] for i in weather_meta_fields]
weather_meta.append('timestamp')
weather_record_path=['weather', 'weatherForecast']

In [117]:
test_weather = pd.json_normalize(
    data=weather_prior_json,
    record_path=weather_record_path,
    meta=weather_meta

)

test_weather.dtypes

daycode                     int64
dayDescription             object
forecastString             object
iconName                   object
summaryDescription         object
temperatureHigh            object
temperatureLow             object
weather.newSnow            object
weather.last24Hours        object
weather.last48Hours        object
weather.last7Days          object
weather.midMountainBase    object
weather.resortID           object
timestamp                  object
dtype: object

In [115]:
test_weather

daycode dayDescription forecastString             iconName  \
0         0          Today                 partly-cloudy-night   
1         0          Today                 partly-cloudy-night   
2         0          Today                 partly-cloudy-night   
3         0          Today                 partly-cloudy-night   
4         0          Today                         clear-night   
5         0          Today                         clear-night   
6         0          Today                         clear-night   
7         0          Today                         clear-night   
8         0          Today                 partly-cloudy-night   
9         0          Today                 partly-cloudy-night   
10        0          Today                             default   
11        0          Today                 partly-cloudy-night   
12        0          Today                 partly-cloudy-night   
13        0          Today                 partly-cloudy-night   
14        0          Today                 partly-cloudy-night   

                                   summaryDescription temperatureHigh  \
0                                       Partly Cloudy              14   
1                                       Mostly Cloudy              14   
2                                       Partly Cloudy               7   
3                                       Partly Cloudy               7   
4                                               Clear              22   
5                                               Clear              29   
6                                               Clear              22   
7                                               Clear              20   
8                                       Mostly Cloudy              35   
9                                       Partly Cloudy              18   
10  Cloudy with periods of light snow developing m...              25   
11                                      Mostly Cloudy              30   
12                                      Partly Cloudy              10   
13                                      Mostly Cloudy              41   
14                                      Mostly Cloudy              35   

   temperatureLow weather.newSnow weather.last24Hours weather.last48Hours  \
0              -6               7                   7                   7   
1              -4               7                   7                   7   
2             -10               3                   5                   5   
3              -8               2                   3                   3   
4               4               0                   0                   2   
5               7               0                   0                   2   
6               6               0                   0                   2   
7               7               7                   7                   7   
8              27               0                   0                   0   
9              -3               0                   1                   4   
10             25               1                   1                   4   
11             20               0                   1                   2   
12             -8               1                   3                   3   
13             26               0                   0                   0   
14             24               0                   0                   0   

   weather.last7Days weather.midMountainBase weather.resortID  \
0                 13                      58                1   
1                 10                      49                2   
2                  8                      52                3   
3                  6                      42                4   
4                  2                      43                5   
5                  2                      18                7   
6                  2                      52                9   
7                 10     

In [23]:
terrain_prior_json

{'timestamp': '2020-02-04 20:37:27.159818-08:00',
 'terrain': [{'runID': 10,
   'resortID': 1,
   'groomed': 'Yes',
   'runName': 'Blue Ox',
   'runType': 3,
   'status': 'O',
   'terrainName': 'Golden Peak'},
  {'runID': 11,
   'resortID': 1,
   'groomed': 'No',
   'runName': 'Boomer',
   'runType': 1,
   'status': 'O',
   'terrainName': 'Golden Peak'},
  {'runID': 12,
   'resortID': 1,
   'groomed': 'No',
   'runName': 'Brisk Walk',
   'runType': 1,
   'status': 'O',
   'terrainName': 'Golden Peak'},
  {'runID': 13,
   'resortID': 1,
   'groomed': 'Yes',
   'runName': 'Choker Cutoff',
   'runType': 2,
   'status': 'O',
   'terrainName': 'Golden Peak'},
  {'runID': 14,
   'resortID': 1,
   'groomed': 'No',
   'runName': 'Fall Line',
   'runType': 3,
   'status': 'O',
   'terrainName': 'Golden Peak'},
  {'runID': 15,
   'resortID': 1,
   'groomed': 'No',
   'runName': 'First Step',
   'runType': 3,
   'status': 'O',
   'terrainName': 'Golden Peak'},
  {'runID': 16,
   'resortID': 1,
  

In [39]:
test_terrain = pd.json_normalize(
    data=terrain_prior_json,
    record_path=['terrain'],
    meta='timestamp'
)

test_terrain.dtypes

runID           int64
resortID        int64
groomed        object
runName        object
runType         int64
status         object
terrainName    object
timestamp      object
dtype: object

In [40]:
test_terrain

runID  resortID groomed         runName  runType status  terrainName  \
0        10         1     Yes         Blue Ox        3      O  Golden Peak   
1        11         1      No          Boomer        1      O  Golden Peak   
2        12         1      No      Brisk Walk        1      O  Golden Peak   
3        13         1     Yes   Choker Cutoff        2      O  Golden Peak   
4        14         1      No       Fall Line        3      O  Golden Peak   
...     ...       ...     ...             ...      ...    ...          ...   
1963   1712        17      No  Double Diamond        4      X    Southface   
1964   1713        17      No    Forrest Bump        4      X    Southface   
1965   1714        17      No    Loose Spruce        4      X    Southface   
1966   1715        17      No     Off The Rim        2      X    Southface   
1967   1716        17      No         Outrage        4      X    Southface   

                             timestamp  
0     2020-02-04 20:37:27.159818-08:00  
1     2020-02-04 20:37:27.159818-08:00  
2     2020-02-04 20:37:27.159818-08:00  
3     2020-02-04 20:37:27.159818-08:00  
4     2020-02-04 20:37:27.159818-08:00  
...                                ...  
1963  2020-02-04 20:37:27.159818-08:00  
1964  2020-02-04 20:37:27.159818-08:00  
1965  2020-02-04 20:37:27.159818-08:00  
1966  2020-02-04 20:37:27.159818-08:00  
1967  2020-02-04 20:37:27.159818-08:00  

[1968 rows x 8 columns]

In [135]:
def jsons_to_df(jsons, record_path, meta='timestamp'):
    if record_path == 'weather':
        # Deal with the nested object that the weather data uses to store the weather forecast
        record_path = weather_record_path
        meta = weather_meta
    df = pd.json_normalize(jsons, record_path=record_path, meta=meta, sep="_")

    # df = set_lifts_df_datatypes(df) update
    return df


def load_merged_json_as_df(merged_file, record_path):
    # load the merged json as a dataframe
    with open(merged_file, "r") as f:
        d = json.load(f)
        df = jsons_to_df(d, record_path)
        return df

In [136]:
jsons_to_df(lifts_prior_json, 'lifts')

liftID  resortID                 liftName status timeToRide  \
0         2         1   Avanti Express Lift #2      O          7   
1         3         1      Wildwood Express #3      O          5   
2         4         1  Mountain Top Express #4      O          4   
3         5         1     High Noon Express #5      O          6   
4         6         1     Riva Bahn Express #6      O          9   
..      ...       ...                      ...    ...        ...   
319      16        17       Green Ridge Triple      X          8   
320       6        17      Orion's Belt Carpet      X          1   
321       8        17              Sachem Quad      X          9   
322       9        17         Snow Star Carpet      X          1   
323      13        17       South Ridge Quad A      X          6   

                            timestamp  
0    2020-02-04 20:37:26.722118-08:00  
1    2020-02-04 20:37:26.722118-08:00  
2    2020-02-04 20:37:26.722118-08:00  
3    2020-02-04 20:37:26.722118-08:00  
4    2020-02-04 20:37:26.722118-08:00  
..                                ...  
319  2020-02-04 20:37:26.722118-08:00  
320  2020-02-04 20:37:26.722118-08:00  
321  2020-02-04 20:37:26.722118-08:00  
322  2020-02-04 20:37:26.722118-08:00  
323  2020-02-04 20:37:26.722118-08:00  

[324 rows x 6 columns]

In [137]:
jsons_to_df(terrain_prior_json, 'terrain')

runID  resortID groomed         runName  runType status  terrainName  \
0        10         1     Yes         Blue Ox        3      O  Golden Peak   
1        11         1      No          Boomer        1      O  Golden Peak   
2        12         1      No      Brisk Walk        1      O  Golden Peak   
3        13         1     Yes   Choker Cutoff        2      O  Golden Peak   
4        14         1      No       Fall Line        3      O  Golden Peak   
...     ...       ...     ...             ...      ...    ...          ...   
1963   1712        17      No  Double Diamond        4      X    Southface   
1964   1713        17      No    Forrest Bump        4      X    Southface   
1965   1714        17      No    Loose Spruce        4      X    Southface   
1966   1715        17      No     Off The Rim        2      X    Southface   
1967   1716        17      No         Outrage        4      X    Southface   

                             timestamp  
0     2020-02-04 20:37:27.159818-08:00  
1     2020-02-04 20:37:27.159818-08:00  
2     2020-02-04 20:37:27.159818-08:00  
3     2020-02-04 20:37:27.159818-08:00  
4     2020-02-04 20:37:27.159818-08:00  
...                                ...  
1963  2020-02-04 20:37:27.159818-08:00  
1964  2020-02-04 20:37:27.159818-08:00  
1965  2020-02-04 20:37:27.159818-08:00  
1966  2020-02-04 20:37:27.159818-08:00  
1967  2020-02-04 20:37:27.159818-08:00  

[1968 rows x 8 columns]

In [138]:
jsons_to_df(weather_prior_json, 'weather')

daycode dayDescription forecastString             iconName  \
0         0          Today                 partly-cloudy-night   
1         0          Today                 partly-cloudy-night   
2         0          Today                 partly-cloudy-night   
3         0          Today                 partly-cloudy-night   
4         0          Today                         clear-night   
5         0          Today                         clear-night   
6         0          Today                         clear-night   
7         0          Today                         clear-night   
8         0          Today                 partly-cloudy-night   
9         0          Today                 partly-cloudy-night   
10        0          Today                             default   
11        0          Today                 partly-cloudy-night   
12        0          Today                 partly-cloudy-night   
13        0          Today                 partly-cloudy-night   
14        0          Today                 partly-cloudy-night   

                                   summaryDescription temperatureHigh  \
0                                       Partly Cloudy              14   
1                                       Mostly Cloudy              14   
2                                       Partly Cloudy               7   
3                                       Partly Cloudy               7   
4                                               Clear              22   
5                                               Clear              29   
6                                               Clear              22   
7                                               Clear              20   
8                                       Mostly Cloudy              35   
9                                       Partly Cloudy              18   
10  Cloudy with periods of light snow developing m...              25   
11                                      Mostly Cloudy              30   
12                                      Partly Cloudy              10   
13                                      Mostly Cloudy              41   
14                                      Mostly Cloudy              35   

   temperatureLow weather_newSnow weather_last24Hours weather_last48Hours  \
0              -6               7                   7                   7   
1              -4               7                   7                   7   
2             -10               3                   5                   5   
3              -8               2                   3                   3   
4               4               0                   0                   2   
5               7               0                   0                   2   
6               6               0                   0                   2   
7               7               7                   7                   7   
8              27               0                   0                   0   
9              -3               0                   1                   4   
10             25               1                   1                   4   
11             20               0                   1                   2   
12             -8               1                   3                   3   
13             26               0                   0                   0   
14             24               0                   0                   0   

   weather_last7Days weather_midMountainBase weather_resortID  \
0                 13                      58                1   
1                 10                      49                2   
2                  8                      52                3   
3                  6                      42                4   
4                  2                      43                5   
5                  2                      18                7   
6                  2                      52                9   
7                 10     

In [32]:
test_terrain.status.unique()

array(['O', 'X'], dtype=object)

In [43]:
??pd.json_normalize

In [206]:
parq_df = load_dataframe_from_parquet_on_s3('lifts' + HISTORY_SUFFIX)

lifts_status_changes_parq_df = get_status_durations(parq_df)
lifts_status_changes_parq_df.to_csv(DATA_DIR + "lifts_status_changes_parq.csv", date_format='%c')
lifts_status_changes_parq_df

liftID resortID                        liftName status  timeToRide  \
0       3       13              7th Heaven Express      O           6   
18      4       13            Jersey Cream Express      O           5   
9       5       13         Excalibur Gondola Lower      O           3   
19      6       13                     Magic Chair      O           6   
15      7       13                 Glacier Express      O           6   
11      8       13             Excelerator Express      O           6   
4       9       13              Catskinner Express      O           4   
7      10       13           Crystal Ridge Express      O           7   
17     11       13                  Horstman T-Bar      X           4   
23     12       13                  Showcase T-Bar      O           3   
24     12       13                  Showcase T-Bar      X           3   
5      14       13             Coca-Cola Tube Park      O           4   
21     22       13             Peak 2 Peak Gondola      O          12   
27     33       13  Whistler Village Gondola Lower      O           5   
6      34       13               Creekside Gondola      O           7   
8      35       13               Emerald 6 Express      O           6   
1      36       13                 Big Red Express      O           8   
16     37       13               Harmony 6 Express      O           6   
20     39       13                   Olympic Chair      O           5   
14     40       13                Garbanzo Express      O           7   
26     41       13                          T-Bars      O           5   
25     42       13                Symphony Express      X           7   
22     43       13                    Peak Express      O           3   
13     44       13                   Franz's Chair      X           8   
12     45       13             Fitzsimmons Express      X           6   
2      69       13         Blackcomb Gondola Lower      O           7   
3      70       13         Blackcomb Gondola Upper      O           7   
10     71       13         Excalibur Gondola Upper      O           5   
28     72       13  Whistler Village Gondola Upper      O          11   

                          timestamp       time_diff  time_diff_seconds  
0  2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
18 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
9  2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
19 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
15 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
11 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
4  2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
7  2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
17 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
23 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
24 2020-02-04 14:48:45.404567-08:00        00:00:00           0.000000  
5  2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
21 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
27 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
6  2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
8  2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
1  2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
16 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
20 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
14 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
26 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
25 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
22 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         596.128553  
13 2020-02-04 14:38:49.276014-08:00 00:09:56.128553         

In [207]:
parq_df.dtypes

liftID                                 category
resortID                               category
liftName                               category
status                                 category
timeToRide                                int64
timestamp     datetime64[ns, America/Vancouver]
dtype: object

In [27]:
s3.Object(BUCKET_NAME, 'lifts' + PRIOR_SUFFIX).delete()
s3.Object(BUCKET_NAME, 'terrain' + PRIOR_SUFFIX).delete()
s3.Object(BUCKET_NAME, 'weather' + PRIOR_SUFFIX).delete()

{'ResponseMetadata': {'RequestId': '6C5D7F609AA102C2',
  'HostId': 'tL+E7xcZ9VDhfz3CQDIZg6E02M83YuQ8iVT4u4XiU9DsGyC2WE0IsblZGJMooBJjnt1LYYleAxU=',
  'HTTPStatusCode': 204,
  'HTTPHeaders': {'x-amz-id-2': 'tL+E7xcZ9VDhfz3CQDIZg6E02M83YuQ8iVT4u4XiU9DsGyC2WE0IsblZGJMooBJjnt1LYYleAxU=',
   'x-amz-request-id': '6C5D7F609AA102C2',
   'date': 'Wed, 05 Feb 2020 04:37:12 GMT',
   'server': 'AmazonS3'},
  'RetryAttempts': 0}}

## Original Version
Handles lifts only

In [47]:
class ParquetWriter():
    """Identifies new data and writes it to Parquet file on S3."""

    def __init__(self):
        # Get current lift status info json
        lifts_current = get_data()['lifts']  # String.
        self.lifts_current_json = json.loads(lifts_current)

        self.lifts_prior_object = s3.Object(BUCKET_NAME, PRIOR_STATUS_FNAME)

    def write_new_data(self):
        """If new data since the last update of Parquet file is found, add it to the Parquet
        file.  Save the current data as json to serve as the prior in the next comparison.
        """

        # Get prior lift status info json
        try:
            self.lifts_prior_object.load()
        except botocore.exceptions.ClientError as e:
            if e.response['Error']['Code'] == "404":
                print("Prior doesn't exist")
                # Create the prior file
                self.save_prior_data()
                print(f"Created {PRIOR_STATUS_FNAME}")
            else:
                # Something else has gone wrong.
                raise
        else:
            # The prior exists
            self.get_prior_data()
            if self.data_changed():

                # Get a df with the status chages between the prior and current json data
                df = jsons_to_df([self.lifts_prior_json, self.lifts_current_json])
                write_dataframe_to_parquet_on_s3(df, HISTORY_FNAME)

                # save current lift status info json as prior
                self.save_prior_data()
                print(
                    f"Replaced data in {self.lifts_prior_object.key} with current data.")

    def get_prior_data(self):
        lifts_prior = self.lifts_prior_object.get()[
            'Body'].read().decode('utf-8')
        self.lifts_prior_json = json.loads(lifts_prior)
        print("Loaded prior json data from S3")

    def data_changed(self):
        """Compare current data json with prior data json without their timestamps.  The timestamps
        on the current json will always be more recent even when none of the lift statuses have changed.
        """
        if self.lifts_prior_json['lifts'] == self.lifts_current_json['lifts']:
            print("No differences between current and prior data were found.")
            return False
        else:
            print("Found differences between current and prior data.")
            return True

    def save_prior_data(self):
        """Save the current data as prior data on S3."""
        bucket.put_object(Key=PRIOR_STATUS_FNAME,
                          Body=bytes(json.dumps(self.lifts_current_json).encode('UTF-8')))

In [117]:
ParquetWriter().write_new_data()

Loaded prior json data from S3
Found differences between current and prior data.
File wb_lifts_history.parquet found.


/Users/paul/anaconda3/lib/python3.7/site-packages/fastparquet/writer.py:655: FutureWarning: RangeIndex._start is deprecated and will be removed in a future version. Use RangeIndex.start instead
  index_cols = [{'name': index_cols.name, 'start': index_cols._start,
/Users/paul/anaconda3/lib/python3.7/site-packages/fastparquet/writer.py:656: FutureWarning: RangeIndex._stop is deprecated and will be removed in a future version. Use RangeIndex.stop instead
  'stop': index_cols._stop, 'step': index_cols._step,
/Users/paul/anaconda3/lib/python3.7/site-packages/fastparquet/writer.py:656: FutureWarning: RangeIndex._step is deprecated and will be removed in a future version. Use RangeIndex.step instead
  'stop': index_cols._stop, 'step': index_cols._step,


Writing 14 records to wb_lifts_history.parquet.
Replaced data in lifts_prior.json with current data.


**Warnings**

See https://github.com/dask/fastparquet/issues/477 for fastparquet warnings about `RangeIndex._start, RangeIndex._stop, RangeIndex._step`


    /Users/paul/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:90: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
    /Users/paul/anaconda3/lib/python3.7/site-packages/fastparquet/writer.py:655: FutureWarning: RangeIndex._start is deprecated and will be removed in a future version. Use RangeIndex.start instead
      index_cols = [{'name': index_cols.name, 'start': index_cols._start,
    /Users/paul/anaconda3/lib/python3.7/site-packages/fastparquet/writer.py:656: FutureWarning: RangeIndex._stop is deprecated and will be removed in a future version. Use RangeIndex.stop instead
      'stop': index_cols._stop, 'step': index_cols._step,
    /Users/paul/anaconda3/lib/python3.7/site-packages/fastparquet/writer.py:656: FutureWarning: RangeIndex._step is deprecated and will be removed in a future version. Use RangeIndex.step instead
      'stop': index_cols._stop, 'step': index_cols._step,

In [182]:
def s3_object_exists(fname):
    """Check if an s3 object exists.  Returns `True` if the object exists."""
    try:
        s3.Object(BUCKET_NAME, fname)
    except botocore.exceptions.ClientError as e:
        if e.response['Error']['Code'] == "404":
            print(f"{fname} doesn't exist")
        else:
            raise
    return True


def load_dataframe_from_parquet_on_s3(fname):
    """ Load a dataframe from a Parquet file on S3. """
    if s3_object_exists(fname):
        read_file = f"s3://{BUCKET_NAME}/{fname}"
        pf = ParquetFile(read_file, open_with=myopen)
        df = pf.to_pandas()

        # Reshift the timezone because parquet engines don't handle shifted timezones
        df.loc[:, 'timestamp'] = df.loc[:, 'timestamp'].dt.tz_convert(TZ)

        return df

In [200]:
parq_df = load_dataframe_from_parquet_on_s3(HISTORY_FNAME)

lifts_status_changes_parq_df = get_status_durations(parq_df)
lifts_status_changes_parq_df.to_csv(DATA_DIR + "lifts_status_changes_parq.csv", date_format='%c')
lifts_status_changes_parq_df

In [202]:
lifts_status_changes_parq_df.sort_values('timestamp')

liftID resortID                 liftName status  timeToRide  \
0       3       13       7th Heaven Express      O           6   
23     12       13           Showcase T-Bar      O           3   
26     41       13                   T-Bars      O           5   
2      69       13  Blackcomb Gondola Lower      O           7   
21     22       13      Peak 2 Peak Gondola      X          12   
..    ...      ...                      ...    ...         ...   
73     42       13         Symphony Express      O           7   
74     45       13      Fitzsimmons Express      X           6   
75     14       13      Coca-Cola Tube Park      O           4   
76     42       13         Symphony Express      X           7   
77     42       13         Symphony Express      X           7   

                          timestamp       time_diff  time_diff_seconds  
0  2020-02-03 14:21:57.064742-08:00 00:47:48.055160        2868.055160  
23 2020-02-03 14:21:57.064742-08:00 00:47:48.055160        2868.055160  
26 2020-02-03 14:21:57.064742-08:00 00:47:48.055160        2868.055160  
2  2020-02-03 14:21:57.064742-08:00 01:15:33.492083        4533.492083  
21 2020-02-03 14:21:57.064742-08:00 18:49:17.304856       67757.304856  
..                              ...             ...                ...  
73 2020-02-04 10:51:07.675750-08:00 03:47:41.600264       13661.600264  
74 2020-02-04 11:11:03.627551-08:00 03:30:00.349858       12600.349858  
75 2020-02-04 12:41:10.947905-08:00 01:59:53.029504        7193.029504  
76 2020-02-04 14:38:49.276014-08:00 00:02:14.701395         134.701395  
77 2020-02-04 14:41:03.977409-08:00        00:00:00           0.000000  

[78 rows x 8 columns]

In [124]:
parq_df.status.cat.categories

Index(['X', 'H', 'O'], dtype='object')

In [125]:
print(*parq_df.status)

O O O O O O O O O O O O X X O O O X O O O X O O O X O O O X X X X X X X X X X X X X X X X X X X X O O O O O O O O O O O O O O O O O O O O O O O O O X O


In [186]:
parq_df.sort_values(["liftName", "timestamp"])

liftID resortID                        liftName status  timeToRide  \
0       3       13              7th Heaven Express      O           6   
29      3       13              7th Heaven Express      X           6   
67      3       13              7th Heaven Express      O           6   
1      36       13                 Big Red Express      O           8   
35     36       13                 Big Red Express      X           8   
..    ...      ...                             ...    ...         ...   
47     33       13  Whistler Village Gondola Lower      X           5   
65     33       13  Whistler Village Gondola Lower      O           5   
28     72       13  Whistler Village Gondola Upper      O          11   
48     72       13  Whistler Village Gondola Upper      X          11   
66     72       13  Whistler Village Gondola Upper      O          11   

                          timestamp  
0  2020-02-03 14:21:57.064742-08:00  
29 2020-02-03 15:09:45.119902-08:00  
67 2020-02-04 09:21:14.155051-08:00  
1  2020-02-03 14:21:57.064742-08:00  
35 2020-02-03 15:37:30.556825-08:00  
..                              ...  
47 2020-02-03 15:37:30.556825-08:00  
65 2020-02-04 09:11:14.369598-08:00  
28 2020-02-03 14:21:57.064742-08:00  
48 2020-02-03 15:37:30.556825-08:00  
66 2020-02-04 09:11:14.369598-08:00  

[76 rows x 6 columns]

### Issue when running get_status_durations(parq_df)
Resulting in error:

    ~/anaconda3/lib/python3.7/site-packages/pandas/core/arrays/categorical.py in from_codes(cls, codes, categories, ordered, dtype)
        705 
        706         if len(codes) and (codes.max() >= len(dtype.categories) or codes.min() < -1):
    --> 707             raise ValueError("codes need to be between -1 and " "len(categories)-1")
        708 
        709         return cls(codes, dtype=dtype, fastpath=True)

    ValueError: codes need to be between -1 and len(categories)-1


Same error seen when running `parq_df[['status']].sort_values(by=['status'])`

This was caused by missing categories (`H`) in the `status` column (and maybe others)

#### Code to inspect issue:

In [228]:
# Test for issue
parq_df[['status']].sort_values(by=['status'])

status
0       H
59      H
58      H
57      H
56      H
..    ...
38      O
39      O
40      O
42      O
65      O

[66 rows x 1 columns]

In [280]:
parq_df.status.cat.categories

Index(['X', 'H', 'O'], dtype='object')

In [281]:
print(*parq_df.status.cat.codes)

2 2 2 2 2 0 2 2 2 2 2 2 2 2 0 2 2 2 0 2 2 2 0 2 0 2 0 2 2 2


In [205]:
len(parq_df.status.cat.codes)

65

In [185]:
# Should be false
parq_df.status.cat.codes.max() >= len(parq_df.status.dtype.categories)

False

In [186]:
# Should be false
parq_df.liftName.cat.codes.min() < -1

False

In [ ]:
for c in parq_df.columns:
    print(parq_df[c].cat.categories)

In [36]:
parq_df.dtypes

liftID                   category
resortID                 category
liftName                 category
status                   category
timeToRide                  int64
timestamp     datetime64[ns, UTC]
dtype: object

In [113]:
lifts_status_changes_df.dtypes

liftID                                      category
resortID                                    category
liftName                                    category
status                                      category
timeToRide                                     int64
timestamp     datetime64[ns, pytz.FixedOffset(-480)]
dtype: object

In [106]:
parq_df["timestamp"] = pd.to_datetime(pd.Series(np.asarray(parq_df["timestamp"])))

In [73]:
read_file = f"s3://{BUCKET_NAME}/{fname}.parquet"
pf = ParquetFile(read_file, open_with=myopen)

# Check the categories for a specific row group
pf.grab_cats(columns='status', row_group_index=1)

{'status': array(['H', 'O', 'X'], dtype=object)}

In [66]:
# If partitioning by column, gives known values for each column
pf.cats

{}

#### Possible solutions
1. Remove partitioning by date column when writing to parquet
2. Set status categories manually via `set_categories`. (and any other columns with the same issue.  See https://github.com/dask/dask/issues/2944
3. Leave problem columns as text-based when writing and loading from parquet

# Testing timestamps for file loading

In [9]:
read_file = f"s3://{BUCKET_NAME}/{fname}.parquet"
pf = ParquetFile(read_file, open_with=myopen)
test = pf.to_pandas()["timestamp"]

In [100]:
# If needed: to convert for categorical datetime to regular datetime
df["timestamp"] = pd.to_datetime(pd.Series(np.asarray(df["timestamp"])))

/Users/paul/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Converting timezone-aware DatetimeArray to timezone-naive ndarray with 'datetime64[ns]' dtype. In the future, this will return an ndarray with 'object' dtype where each element is a 'pandas.Timestamp' with the correct 'tz'.
	To accept the future behavior, pass 'dtype=object'.
	To keep the old behavior, pass 'dtype="datetime64[ns]"'.
  


In [101]:
test.dt = test.dt.tz_convert(tz= 'America/Vancouver')

/Users/paul/anaconda3/lib/python3.7/site-packages/pandas/core/series.py:597: FutureWarning: Converting timezone-aware DatetimeArray to timezone-naive ndarray with 'datetime64[ns]' dtype. In the future, this will return an ndarray with 'object' dtype where each element is a 'pandas.Timestamp' with the correct 'tz'.
	To accept the future behavior, pass 'dtype=object'.
	To keep the old behavior, pass 'dtype="datetime64[ns]"'.


more info: https://pandas-docs.github.io/pandas-docs-travis/whatsnew/v0.24.0.html#converting-timezone-aware-series-and-index-to-numpy-arrays

In [102]:
load_dataframe_from_parquet_on_s3(fname).dtypes

liftID        category
resortID      category
liftName      category
status        category
timeToRide       int64
timestamp     category
dtype: object

### Testing local parquet saves

In [ ]:
def save_parquet(df, fname):
    # parquet engines don't handle shifted timezones
    df.loc[:, 'timestamp'] = df.loc[:, 'timestamp'].dt.tz_convert(pytz.utc)

    # Note: May need snappy-python as a req to run on AWS Lambda
    df.to_parquet(DATA_DIR + fname + '.parquet',
                  engine='fastparquet',
                  partition_on=['timestamp'],
                  file_scheme='mixed')

In [91]:
save_parquet(df[0:3].copy(), 'wb_lifts_history')

In [92]:
df.iloc[20:22, :].copy().to_parquet(DATA_DIR + 'wb_lifts_history' + '.parquet',
              engine='fastparquet',
              partition_on=['timestamp'],
              file_scheme='mixed',
              append=True)
# Catch exception that is doesn't exist here

In [ ]:
# todo: change time_diff to "duration"
# test on lambda
# make datatype dict for and general set datatypes function


# Utilities

In [118]:
def del_lifts_history():
    s3.Object(BUCKET_NAME, HISTORY_FNAME).delete()

def del_lifts_prior():
    s3.Object(BUCKET_NAME, PRIOR_STATUS_FNAME).delete()

In [121]:
del_lifts_history()
del_lifts_prior()